EIPS/eip-7791.md

```diff
@@ -0,0 +1,82 @@
+This increased demand also benefits miners and validators, who are rewarded for executing these contracts.
+
+Monetizing smart contracts in a scalable manner remains challenging at the time of this writing.
+This difficulty is evident from existence of many different monetization strategies employed across various smart contracts — ranging from fee structures to the issuance of tokens with "tokenomics" of varying levels of complexity.
+Additionally, many public goods projects struggle to secure funding.
+
+Introducing the `GAS2ETH` opcode offers contract authors, as well as the tools they use, a new way to achieve their monetization objectives.
+By charging gas, they integrate with an established user experience that is both familiar and understood by users.
+The proposed instruction ensures that existing transaction creation and processing tools remain unchanged.
+Moreover, by charging gas, contract authors align economically with network activity; they benefit from higher compensation during periods of intense network usage and receive less when activity is low.
+This helps align the incentives of smart contract authors, validators, and the broader network.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+A new opcode is introduced, `GAS2ETH` (`0xFC`), which:
+
+- Pops two values from the stack: `addr` then `gas_amount`. If there are fewer than two values on the stack, the calling context should fail with stack underflow.
+- Deducts `gas_amount` from the current calling context.
+- Computes `wei_val` by multiplying `gas_amount` by the current transaction context's `gas_price`.
+- Endows the address `addr` with `wei_val` wei.
+- If the gas cost of this opcode + `gas_amount` is greater than the available gas in the current calling context, the calling context should fail with "out-of-gas" and any state changes reverted.
+- Pushes `wei_val` onto the stack.
+
+Note that the transfer of `wei_val` to the given account cannot fail. In particular, the destination account code (if any) is not executed, or, if the account does not exist, the balance is still added to the given address `addr`.
+
+The proposed cost of this opcode is similar to the recently proposed `PAY` opcode, but changing the base cost from `9000` to `2400`. That is:
+
+- The base cost of this opcode is `2400`. This is priced so that invoking `GAS2ETH` on a cold account costs the same as a cold `SSTORE`.
+- If `addr` is not the zero address, the [EIP-2929](./eip-2929.md) account access costs for `addr` (but NOT the current account) are also incurred: 100 gas for a warm account, 2600 gas for a cold account, and 25000 gas for a new account.
```

Comment:
What does the zero address have to do with this? Does this change some behavior of this opcode?

Reply:
that must be a mistake that slipped in over the course of revisions, will fix